src/backend/distributed/planner/multi_explain.c

```diff
@@ -99,10 +173,655 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+{
+	DestReceiver *dest = NULL;
+	int eflags = 0;
+	QueryDesc *queryDesc = NULL;
+	int instrument_option = 0;
+
+	if (es->analyze && es->timing)
+	{
+		instrument_option |= INSTRUMENT_TIMER;
+	}
+	else if (es->analyze)
+	{
+		instrument_option |= INSTRUMENT_ROWS;
+	}
+
+	if (es->buffers)
+	{
+		instrument_option |= INSTRUMENT_BUFFERS;
+	}
+
+	/*
+	 * Use a snapshot with an updated command ID to ensure this query sees
+	 * results of any previously executed queries.
+	 */
+	PushCopiedSnapshot(GetActiveSnapshot());
+	UpdateActiveSnapshotCommandId();
+
+	/*
+	 * Normally we discard the query's output, but if explaining CREATE TABLE
+	 * AS, we'd better use the appropriate tuple receiver.
+	 */
+	if (into)
+	{
+		dest = CreateIntoRelDestReceiver(into);
+	}
+	else
+	{
+		dest = None_Receiver;
+	}
+
+	/* Create a QueryDesc for the query */
+	queryDesc = CreateQueryDesc(masterPlan, queryString,
+								GetActiveSnapshot(), InvalidSnapshot,
+								dest, params, instrument_option);
+
+	/* Select execution options */
+	if (es->analyze)
+	{
+		eflags = 0;             /* default run-to-completion flags */
+	}
+	else
+	{
+		eflags = EXEC_FLAG_EXPLAIN_ONLY;
+	}
+	if (into)
+	{
+		eflags |= GetIntoRelEFlags(into);
+	}
+
+	/*
+	 * ExecutorStart creates the merge table. If using ANALYZE, it also executes the
+	 * worker job and populates the merge table.
+	 */
+	ExecutorStart(queryDesc, eflags);
+
+	if (es->analyze)
+	{
+		ScanDirection dir;
+
+		/* if using analyze, then finish query execution */
+
+		/* EXPLAIN ANALYZE CREATE TABLE AS WITH NO DATA is weird */
+		if (into && into->skipData)
+		{
+			dir = NoMovementScanDirection;
+		}
+		else
+		{
+			dir = ForwardScanDirection;
+		}
+
+		/* run the plan */
+		ExecutorRun(queryDesc, dir, 0L);
+
+		/* run cleanup too */
+		ExecutorFinish(queryDesc);
+	}
+
+	/*
+	 * ExplainOnePlan executes the master query again, which ensures that the execution
+	 * time only shows the execution time of the master query itself, instead of the
+	 * overall execution time.
+	 */
+	ExplainOnePlan(queryDesc->plannedstmt, into, es, queryString, params, planDuration);
+
+	/*
+	 * ExecutorEnd for the distributed query is deferred until after the master query
+	 * is executed again, otherwise the merge table would be dropped.
+	 */
+	ExecutorEnd(queryDesc);
+
+	FreeQueryDesc(queryDesc);
+
+	PopActiveSnapshot();
+}
+
+
+/*
+ * ExplainJob shows the EXPLAIN output for a Job in the physical plan of
+ * a distributed query by showing the first remote EXPLAIN for the first
+ * task.
+ */
+static void
+ExplainJob(Job *job, ExplainState *es)
+{
+	List *taskList = job->taskList;
+	ListCell *taskCell = NULL;
+	int taskCount = list_length(taskList);
+
+	ExplainOpenGroup("Job", NULL, true, es);
+
+	ExplainPropertyInteger("Task Count", taskCount, es);
+
+	if (ExplainAllTasks || taskCount <= 1)
+	{
+		ExplainPropertyText("Tasks Shown", "All", es);
+	}
+	else
+	{
+		ExplainPropertyText("Tasks Shown", "Sample", es);
```

Comment:
`Sample` isn't really accurate - isn't this most of the the time going to be deterministic task?